### PR TITLE
fix: add api key support

### PR DESCRIPTION
**Pull request by seveibar**
* adds api key support
* is `seam_seedkey1_seedkey1` a good/memorable api key?

### CODE REVIEW

File: src/lib/database/seed.ts [hunk 1]
```diff
@@ -8,9 +8,11 @@ export interface Seed {
   august_device_1: "august_device_1"
   august_device_2: "august_device_2"
   schlage_device_1: "schlage_device_id"
+  seam_seedkey1_seedkey1: "seam_seedkey1_seedkey1"
+  seam_seedkey2_seedkey2: "seam_seedkey2_seedkey2"
 }
 
-export const seed = (db: Database): Seed => {
+export const seed = (db: DatabaseState & DatabaseMethods): Seed => {
```
Comment (razor-x) [inline src/lib/database/seed.ts:15]: Odd, `Database` should be correct here. Can I ask what the type error was?

File: test/fixtures/get-test-server.ts [hunk 2]
```diff
@@ -25,9 +25,10 @@ interface ApiRequest extends NextApiRequest {
 }
 
 export const getTestServer = async (
-  t: ExecutionContext
+  t: ExecutionContext,
+  { seed = true }: { seed?: boolean } = {}
```
Comment (razor-x) [inline test/fixtures/get-test-server.ts:29]: Why was this added? Is there a use case for testing with an unseeded DB? I would prefer a separate function `getUnseededTestServer` in that case (then you can avoid the type issues 😄).

**Comment by razor-x** [comment]
Is there a reason the name has duplicated parts, or is that a typo? `seam_seedkey1_seedkey1`

File: test/fixtures/get-test-server.ts (hunk 2 above)
Comment (seveibar) [reply test/fixtures/get-test-server.ts:29]: there's supported to be a rule around only having one getTestServer for london-style testing so that the test always look the same, this is basically a seam connect convention here to provide the seed parameter. Yea the type issue is annoying but I feel like it's just a temporary intentional bug since typing this function is possible, I'm just trying to not get fancy atm

**Comment by seveibar** [comment]
> Is there a reason the name has duplicated parts, or is that a typo? seam_seedkey1_seedkey1

@razor-x this is intentional to try to make it easy to remember. Do you have an idea on an easier-to-remember name? `seam_seedkey1_0000` or something?

File: src/lib/database/seed.ts (hunk 1 above)
Comment (seveibar) [reply src/lib/database/seed.ts:15]: the db passed into seed sometimes doesn't have `getState` and `setState` if it comes from the fake IIRC. This is a bit narrower so IMO it's ok, broadening types is generally worrying

File: test/fixtures/get-test-server.ts (hunk 2 above)
Comment (seveibar) [reply test/fixtures/get-test-server.ts:29]: I can fix the type issue if you think it's worth blocking

File: src/lib/database/seed.ts (hunk 1 above)
Comment (razor-x) [reply src/lib/database/seed.ts:15]: Ah, I see! Yea actually we should probably just update `Database` to be the same as here: https://github.com/seamapi/fake-seam-connect/blob/60f9a0548f397ed85c52d24a091638c8803d25c1/src/lib/fake.ts#L11

That update I made felt like a hack. If we need to use the object with the zustand methods it should be an internal type (ZustandDatabase or Store?).

File: test/fixtures/get-test-server.ts (hunk 2 above)
Comment (razor-x) [reply test/fixtures/get-test-server.ts:29]: Many of the changes we are adding here might be back-ported to the template, that is why I'm flagging things like this. I guess I still am unsure what is going to use this param. 

If you have a better solution for the types, that would be nice. Not understanding the use case is what is making it hard for me to make a call. If people will write test with this, I think they should be confident in the types and not get null errors.

Comment (seveibar) [reply test/fixtures/get-test-server.ts:29]: I think sometimes it is nice to have a blank fake, especially when doing security-access testing

File: src/lib/database/schema.ts [hunk 3]
```diff
@@ -29,6 +31,11 @@ export interface DatabaseMethods {
     created_at?: string
     workspace_id?: string
   }) => Workspace
+  addAPIKey: (params: {
```
Comment (razor-x) [inline src/lib/database/schema.ts:34]: It has to be like this since this is our name for `api_key` not our `api_k_e_y`: otherwise we break snake case to param case transformations (if you transform then transform back you get a different result).

```suggestion
  addApiKey: (params: {
```

File: src/lib/database/schema.ts [hunk 4]
```diff
@@ -29,6 +31,11 @@ export interface DatabaseMethods {
...
+    name?: string
+    token?: string
+    created_at?: string
+  }) => APIKey
```
Comment (razor-x) [inline src/lib/database/schema.ts:38]:
```suggestion
  }) => ApiKey
```

File: src/lib/database/schema.ts (hunk 3 above)
Comment (seveibar) [reply src/lib/database/schema.ts:34]: I'm ok with establishing that as the Seam convention, but we have historically used airbnb convention https://twitter.com/seveibar/status/1585389759303938048?s=20

Comment (seveibar) [reply src/lib/database/schema.ts:34]: Airbnb convention in case you aren't on twitter haha

image

**Comment by razor-x** [comment]
I don't think it can be `seam_seedkey1_seedkey1` since this looks like a mistake. It can't be `seed_api_key_1` since that would be its id. These keys are not for a specific workspace, otherwise I'd suggest `seed_workspace_1_api_key`.

So, maybe `seed_api_key_1_token`?

File: src/lib/database/schema.ts (hunk 3 above)
Comment (seveibar) [reply src/lib/database/schema.ts:34]: Also note that microsoft has a different convention for two letter acronyms. The convention you're suggesting is the Google convention I think, which I'm fine with since our internal usage of snake case makes anything else pretty annoying, and I'm a bit indifferent

Comment (seveibar) [reply src/lib/database/schema.ts:34]: Also I guess I should also say I should probably respect the convention of the project lol- so not super relevant, but always fun to talk about camel case

**Comment by seveibar** [comment]
> I don't think it can be `seam_seedkey1_seedkey1` since this looks like a mistake. It can't be `seed_api_key_1` since that would be its id. These keys are not for a specific workspace, otherwise I'd suggest `seed_workspace_1_api_key`.
> 
> So, maybe `seed_api_key_1_token`? It is the token after all and not the api key object or id. Or `seed_api_token_1`?

I don't think we should have more than 3 underscore terms, because imo that looks structurally different and might be parsed in the future differently. It has to start with `seam_`, we could do `seam_seed_key1`?

File: src/lib/database/schema.ts (hunk 3 above)
Comment (razor-x) [reply src/lib/database/schema.ts:34]: Yes, but IMO there are two competing conventions. With `HTTPRequest` there is no `http_request` or `seam.httpRequest` method. It's just the name of the idea. 

Since Seam is highly concerned with supporting the SDK in various languages, I think a stronger convention is the unambiguous transformation between underscore and camel case. At least from my POV, it is frustrating having to guess what _might_ be an acronym or not (actually Initialism).

This thread may not be the best forum for this particular discussion though. Fun fact, golang basically forces you to use this convention for `ID` https://github.com/golang/go/wiki/CodeReviewComments#initialisms. If we followed this convention consistently, then we would always write `fooID`. If you only live in Go, then you don't really notice this too much (unless on the edges of api boundaries that use `fooId`).

**Comment by razor-x** [comment]
> I don't think we should have more than 3 underscore terms
👍🏻 

Does it need seam or seed prefixes. I don't love `key` by itself. I'd expect `apikey` or `apitoken` or `token`. 

- `seam_api_token1`
- `seed_api_token1`
- `seam_seed_apitoken1`
- `seam_seed_token1`

**Comment by seveibar** [comment]
Yea I'm also thinking we might need to make the short prefix unique across the api keys (just realized this- we don't allow short token duplication within a single workspace)

* `seam_apikey1_0000`
* `seam_apikey1_token`

wdyt- didn't think this would be so annoying

**Comment by razor-x** [comment]
Ah right they have to have the proper format! Either of those last two suggestions look fine! `seam_apikey1_token` is a bit nicer to read / talk about, etc.

File: src/lib/database/schema.ts (hunk 4 above)
Comment (seveibar) [reply src/lib/database/schema.ts]: 6f1dbf5

File: src/lib/database/schema.ts (hunk 3 above)
Comment (seveibar) [reply src/lib/database/schema.ts]: 6f1dbf5

File: src/lib/database/seed.ts (hunk 1 above)
Comment (seveibar) [reply src/lib/database/seed.ts]: 544eb50

File: test/fixtures/get-test-server.ts (hunk 2 above)
Comment (seveibar) [reply test/fixtures/get-test-server.ts]: imo too complex (it might make sense if it was outside a test suite to get complex, but idk) 5593413

**Review by razor-x — APPROVED** [review]
Database vs ZustandDatabase looks much better. Will backport to the template.

I agree the the types for the seedless db are a bit too complicated now and they still didn't remove the any cast. We can keep this for now, but I might think about what is going on here a bit more and why the type
system is fighting this pattern vs supporting us here.

**Comment by seveibar** [comment]
> I agree the the types for the seedless db are a bit too complicated now and they still didn't remove the any cast. We can keep this for now, but I might think about what is going on here a bit more and why the type
system is fighting this pattern vs supporting us here.

@razor-x FWIW the internal any-cast doesn't really matter, the new types made the type that the consumer sees correct so it is an improvement. I also think internal any-casting is a really good practice when the type system prevents internal-facing code from being simple.

**Comment by seveibar** [comment]
Also I merged this as a fix because I didn't see we were in v1, sorry about that....
